packages/backend/src/peripherals/ethereum/EthereumClient.ts

```diff
@@ -79,13 +80,66 @@ export class EthereumClient {
     const logs: providers.Log[] = []
     for (const batch of batches) {
       const nestedLogs = await Promise.all(
-        batch.map((filter) => this.provider.getLogs(filter))
+        batch.map((filter) => {
+          if (
+            'fromBlock' in filter &&
+            typeof filter.address === 'string' &&
+            typeof filter.fromBlock === 'number' &&
+            typeof filter.toBlock === 'number'
+          ) {
+            return this.getAllLogs(
+              filter.address,
+              filter.topics,
+              filter.fromBlock,
+              filter.toBlock
+            )
+          }
+          return this.provider.getLogs(filter)
+        })
       )
       logs.push(...nestedLogs.flat())
     }
     return logs
   }
 
+  async getAllLogs(
```

Comment:
It would be useful to test this function, especially the binary search part.

Reply:
ah yes. I've copied it straight from L2BEAT's repository, I'll just copy the tests too